llama.cpp

```diff
@@ -10610,6 +10709,178 @@ struct llm_build_context {
+                    freq_base, freq_scale, ext_factor, attn_factor, beta_fast, beta_slow
+                );
+
+                // So because our original wo matrix wasn't 3x, the below function fails because there aren't enough elems in it.
+                // Got: [head_dim][n_tokens][n_head_v]
+                // Want: [n_embd_v_gqa(384)][n_tokens]
+                // I guess this means that i need to be able to able to repeat them
+                // Assertion failed: (v_cur->ne[0] == n_embd_v_gqa && v_cur->ne[1] == n_tokens), function llm_build_kv_store, file llama.cpp, line 6309.
+                // In the python version it does this:
+                /*
+                if self.num_groups != 1:
+                        # GQA
+                        # [B, k_h, S, h] --> [B, q_h, S, h] // so, k=3 -> q=12
+                        keys = keys.repeat_interleave(self.num_groups, dim=1)
+                        # [B, v_h, S, h] --> [B, q_h, S, h] // so, v=3 -> q=12
+                        values = values.repeat_interleave(self.num_groups, dim=1)
+
+                ...
+
+                attn_output = F.scaled_dot_product_attention(
+                    queries,
+                    keys,
+                    values,
+                    attn_mask=causal_mask,
+                    dropout_p=0,
+                )
+
+                attn_output = attn_output.transpose(1, 2).contiguous()
+                attn_output = attn_output.reshape(
+                    batch_size, seq_length, self.num_q_heads * self.head_dim
+                )
+                attn_output = self.out_proj(attn_output)
+                if not output_attentions:
+                    attn_weights = None
+                return attn_output, attn_weights, past_key_value
+                 *
+                 */
+                cb(Kcur, "Kcur", il);
+                cur = llm_build_kv(ctx0, model, modified_hparams, kv_self, gf,
+                     model.layers[il].wo, NULL,
+                    Kcur, Vcur, Qcur, KQ_mask, nullptr, n_ctx, n_tokens/2, kv_head, n_kv, 1.0f, cb, il);
+            }
+            if (il == n_layer - 1) {
+                // skip computing output for unused tokens
+                struct ggml_tensor* inp_out_ids = build_inp_out_ids();
+                cur = ggml_get_rows(ctx0, cur, inp_out_ids);
+                residual = ggml_get_rows(ctx0, residual, inp_out_ids);
+            }
+            cur = ggml_add(ctx0, cur, residual);
+            residual = cur;
+            cur = llm_build_norm(ctx0, cur, modified_hparams,
+                model.layers[il].ffn_norm, NULL,
+                LLM_NORM_RMS, cb, il);
+            cb(cur, "ffn_norm", il);
+            // FF
+            {
+
+                cur = llm_build_norm(ctx0, cur, hparams,
+                model.layers[il].ffn_norm, NULL,
+                LLM_NORM_RMS, cb, il);
+                cb(cur, "ffn_norm", il);
+
+                cur = llm_build_ffn(ctx0, cur,
```

Comment:
Now i think i just need to implement `LLM_FFN_SWIGLU` which is slightly different to the existing activation functions i think:
<img width="1109" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/32605850/7ae61692-2fdf-4305-b254-ec952a356042">